launch/launch/descriptions/executable.py

```diff
@@ -0,0 +1,141 @@
+"""Module for a description of an Executable."""
+
+import os
+import shlex
+import threading
+from typing import Dict
+from typing import Iterable
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+from launch.some_substitutions_type import SomeSubstitutionsType
+from launch.substitution import Substitution
+from launch.launch_context import LaunchContext
+from launch.utilities import normalize_to_list_of_substitutions
+from launch.utilities import perform_substitutions
+
+_global_process_counter_lock = threading.Lock()
+_global_process_counter = 0  # in Python3, this number is unbounded (no rollover)
+
+class Executable:
+    """Describes an executable (typically a single process) which may be run by the launch system."""
+
+    def __init__(
+        self, *,
+        cmd: Iterable[SomeSubstitutionsType],
```

Comment:
A bit unrelated: maybe we should take advantage of this new description to fix:
- https://github.com/ros2/launch/issues/263
- https://github.com/ros2/launch/issues/394
?

Using a single string instead than a list of strings plays better with substitutions.

Reply:
I've taken a stab at addressing this. Also added a few unit test cases to evaluate how it's handling the different cases described in #263 itself. I'd very much like to add more tests to make sure we're getting the expected results, so if you have any additional suggestions, I'd love to hear them.